### DISCUSSION TITLE
Provide example of how to test dependencies which depend on Request object

### DISCUSSION THREAD

**Discussion by ascended121** [fastapi/fastapi#6808] [Questions] [answered]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Example

Here's a self-contained, [minimal, reproducible, example](https://stackoverflow.com/help/minimal-reproducible-example) with my use case:

```Python
from fastapi import FastAPI

app = FastAPI()

class User:
    def __init__(self, name):
        self.user = name

def get_user(request: Request):
   # look at request headers to determine user and do other stuff
   return User(request.headers["user"])
   
@app.get("/")
def read_root(user: str = Depends(get_user)):
    return {"Hello": user}
```

### Description

What is the recommended way to test a dependency, like `get_user` in the example above? I cannot find an easy way to instantiate a `Request` object and populate the proper HTTP headers to pass into `get_user` during a test.

I am aware of `dependency_overrides`, however this doesn't help because I don't want to stub out `get_user`, I want to verify that it returns the proper values, which are not easily observable from the responses from the actual endpoints.

I am aware of `TestClient`, however since this app uses a database, it requires significant effort to stub/mock out all the interactions for the entire app, just to exercise this internal method.

### Environment

* OS: Ubuntu 20
* FastAPI Version: 0.63.0
* Python version: 3.8.9

**Comment by SnkSynthesis** [comment, marked as answer]
@ascended121,

You could do:
```python
def test_get_user():
    r = Request(scope={"type": "http", "headers": [[b"user", b"example_username"]]})
    user = get_user(r)
    assert user.user == "example_username"
```

**Comment by ascended121** [comment]
Thank you. I was struggling to figure out how to instantiate the Starlette Request object which was making it seem like it there wasn't an easy way to instaniate them... didn't pick up that it needed to be a double-list containing byte-strings. Much appreciated.

**Comment by tiangolo** [comment]
Thanks for the help here @SnkSynthesis ! 👏  🙇 

Thanks for reporting back and closing the issue @ascended121  👍

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
